### ISSUE TITLE
Are projects with submodules contemplated?

### ISSUE THREAD

**Issue by marcbc** [webpro/release-it#149]
I'm currently working in a multi-repository setup, were I have a parent project which points to several git submodules. Something like:

```
PARENT_REPO
 └──  BACKOFFICE_API_REPO
 └──  BACKOFFICE_UI_REPO
 └──  CLIENT_APP_REPO
 └──  CLIENT_UI_REPO
```

I'd like to use **release-it** to manage the parent project globally as well as the submodules individually.

Currently, I'm having an issue when dealing with the submodules, because one of the initial checks the tool does in **tasks.js** is:
```
if (!await git.isGitRepo()) {
  throw new Error('Not a git repository, Marc.');
}
```
which internally, in **git.js** checks for...
```
export function isGitRepo() {
  return run('!git rev-parse --git-dir', { isReadOnly: true }).then(stdout => stdout === '.git', () => false);
}
```

Executing this `git rev-parse` in the `PARENT_REPO` path, does indeed return the git directory `.git`.

But when executing it in the submodules, the git data belongs to the parent project, and the path then is relative to it (something like `/PARENT_REPO/.git/modules/BACKOFFICE_API_REPO`, not passing the `=== '.git'` test.

I have tried to disable the check for this case, and everything seems to be working fine.

My question is if this use case is contemplated in some other way that I'm missing. If not do you think it would be useful considering it for future releases?

thanks!

**Comment by webpro** [comment]
For starters, in "regular" repos (say, without submodules), release-it should be run from the root dir, as all sorts of things could fail (such as build commands, finding the dist dir, etc).

On to your specific issue. You are running `release-it` from inside the `BACKOFFICE_API_REPO` directory? And that submodule does have a `.git` directory, or not?

Btw, I like personalized error messages, too :-)

**Comment by marcbc** [comment]
haha, I forgot I updated that message... :)

I'm running release-it form the submodule directories yes, since I want to be able to release versions of the submodules, independently (e.g. backoffice can be in 1.0.10 and client in 1.0.19) and from time to time, I'm releasing the whole project from the root.

It seems that by design, submodules keep their .git data in the "parent project", so there is no `.git` directory, which is what is causing release-it to fail.

So the problem I'm facing is that the root dir of my projects doesn't always contain the .git dir.

Maybe my setup is convoluted and a very strange edge case, if you think so just forget about it. 

thanks!

**Comment by webpro** [comment]
I will give it another thought later on. If it wouldn't be too dirty, we might add an option or something to disable or alter that `isGitRepo` check.

**Comment by vchagaev** [comment]
I have the same issue.

**Comment by webpro** [comment]
Sounds like a legit use case. Not sure how this works out, but I've released [4.5.0-alpha.0](https://github.com/webpro/release-it/releases/tag/4.5.0-alpha.0) which should allow you to run release-it from inside a Git submodule. You can install with `npm install release-it@alpha (--global)`.

Not sure how things will hold in various circumstances (e.g. to which origin it will push). Please let me know how it goes.

At the very least I've learned about a new, very long argument: [--show-superproject-working-tree](https://git-scm.com/docs/git-rev-parse#git-rev-parse---show-superproject-working-tree)!

**Comment by vchagaev** [comment]
Oh, I'm sorry but my project does not have submodules. I just have several package.json files.
For example release-it from SUB_FOLDER does not work (ERROR Not a git repository.)
I will try to fix it.

ROOT_FOLDER
 └──  SUB_FOLDER
               └── package.json
 └──  package.json

**Comment by vchagaev** [comment]
I mean that checking `stdout === '.git'` is overkill. If command `git rev-parse --git-dir` has no error then it is a git repo. 
It would be cool if you fix that. Everything else works fine.  Thanks!

**Comment by vchagaev** [comment]
https://github.com/webpro/release-it/pull/154

**Comment by webpro** [comment]
The version I've released is a more defensive approach for @marcbc's use case this ticket was opened for. I think it's still valid.

See @vchagaev's PR for a different approach + discussion.

**Comment by marcbc** [comment]
4.5.0-alpha.0 worked for my use case @webpro , thanks for the quick implementation!

**Comment by webpro** [comment]
@marcbc I've used a different approach that should still work in [5.0.0-beta.0](https://github.com/webpro/release-it/releases/tag/5.0.0-beta.0). Hope we can move forward with this v5.

**Comment by marcbc** [comment]
@webpro, I was about to try this 5.0.0-beta.0, but I ran into an issue. 

It seems you are dropping support for NodeJS 4 in this one, but should NodeJS 6 still be supported? I see you're using async which is not available in NodeJS 6.and it's not running in my environment. Should there be some polyfill in place?

**Comment by webpro** [comment]
It should still run in Node v6. Did you see errors during installation/running it, or by reading the notes?

The CLI is using a Babel transpilation step to support e.g. `async/await` syntax.

**Comment by marcbc** [comment]
thanks for the quick answer. I was trying to run it:

```
/../.nvm/versions/node/v6.11.0/lib/node_modules/release-it/lib/tasks.js:16
module.exports = async options => {
                       ^^^^^^^
SyntaxError: Unexpected identifier
    at createScript (vm.js:56:10)
    at Object.runInThisContext (vm.js:97:10)
    at Module._compile (module.js:542:28)
    at Module._extensions..js (module.js:579:10)
    at Object.require.extensions.(anonymous function) [as .js] (/Users/marc.baiges/.nvm/versions/node/v6.11.0/lib/node_modules/release-it/node_modules/babel-register/lib/node.js:152:7)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
    at Module.require (module.js:497:17)
    at require (internal/module.js:20:19)

```

**Comment by webpro** [comment]
Thanks for the catch! Some magic was removed during the rewrite. Should be fixed in [5.0.0-beta.1](https://github.com/webpro/release-it/releases/tag/5.0.0-beta.1).

Just not sure why the tests succeed in Node v6, though.
